gemelli/matrix_completion.py

```diff
@@ -0,0 +1,169 @@
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import numpy as np
+from gemelli.optspace import OptSpace
+from .base import _BaseImpute
+from scipy.spatial import distance
+
+
+class MatrixCompletion(_BaseImpute):
+
+    def __init__(self, n_components=2, max_iterations=5, tol=1e-5):
+        """
+
+        This form of matrix completion uses OptSpace (1). Furthermore,
+        here we directly interpret the loadings generated from matrix
+        completion as a dimensionality reduction.
+
+        Parameters
+        ----------
+
+        X: numpy.ndarray - a rclr preprocessed matrix of shape (M,N)
+        N = Features (i.e. OTUs, metabolites)
+        M = Samples
+
+        n_components: int, optional : Default is 2
+        The underlying rank of the default set
+        to 2 as the default to prevent overfitting.
+
+        max_iterations: int, optional : Default is 5
+        The number of convex iterations to optimize the solution
+        If iteration is not specified, then the default iteration is 5.
+        Which redcues to a satisfactory error threshold.
+
+        tol: float, optional : Default is 1e-5
+        Error reduction break, if the error reduced is
+        less than this value it will return the solution
+
+        Returns
+        -------
+        U: numpy.ndarray - "Sample Loadings" or the unitary matrix
+        having left singular vectors as columns. Of shape (M,n_components)
+
+        s: numpy.ndarray - The singular values,
+        sorted in non-increasing order. Of shape (n_components,n_components).
+
+        V: numpy.ndarray - "Feature Loadings" or Unitary matrix
+        having right singular vectors as rows. Of shape (N,n_components)
+
+        solution: numpy.ndarray - (U*S*V.transpose()) of shape (M,N)
+
+        distance: numpy.ndarray - Distance between each
+        pair of the two collections of inputs. Of shape (M,M)
+
+        Raises
+        ------
+        ValueError
+
+        ValueError
+            `ValueError: n_components must be at least 2`.
+
+        ValueError
+            `ValueError: max_iterations must be at least 1`.
+
+        ValueError
+            `ValueError: Data-table contains either np.inf or -np.inf`.
+
+        ValueError
+            `ValueError: The n_components must be less
+             than the minimum shape of the input table`.
+
+        References
+        ----------
+        .. [1] Keshavan RH, Oh S, Montanari A. 2009. Matrix completion
+                from a few entries (2009_ IEEE International
+                Symposium on Information Theory
+
+        Examples
+        --------
+        TODO
+
+        """
+
+        self.n_components = n_components
+        self.max_iterations = max_iterations
+        self.tol = tol
+
+        return
+
+    def fit(self, X):
```

Comment:
Sklearn transformers `fit`, and `fit_transform` methods typically take an optional `y=None` argument (even if `y` is not used in the computation) [[see here](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html)]. Including this argument makes the logic around chaining together multiple models (e.g., fitting deicode, then using it's components in an ML) a lot easier to use, especially if using some of sklearn's super useful pipeline functionality.

Reply:
y=None is now added. Thanks!